go-controller/pkg/util/pod_annotation.go

```diff
@@ -66,6 +66,9 @@ type PodAnnotation struct {
 	Gateways []net.IP
 	// Routes are additional routes to add to the pod's network namespace
 	Routes []PodRoute
+	// SkipIPConfig if set to true will skip the pod's interface ip
+	// configuration
+	SkipIPConfig *bool
```

Comment:
Propepertly manage "omitempty" with json golang library a the Ovn pod annotation, without the pointer omitempty means "false" since it's the default at golang, but I prefer to have it explicilty set true or false.

Reply:
I am assuming this is going to be set by kubevirt on creation since I don't see anywhere in this PR where it is being set.

If that is the case, I can't say I like it very much:

- The owner of this information up until now has been ovn-k. This is a change of paradigm where we now have two producers for the pod annotation information and a chance of life-cycle for the annotation as well. Not pinpointing any specific issues here, but such changes generally bring problems about.
- If for the most part, this has a most frequently default where it is going to apply to all networks, it can't be optimally expressed in this annotation, where we would be forced to have a value for all networks.
- Most importantly, when we move all this information to an ovn-k CRD I guess we will have to find a new home for this setting anyway as I guess we won't be giving permissions to kubevirt to write to ovn-k CRD.

Reply:
For clarification, meant to say I think this should live on its own annotation.

Reply:
The inital approach from the enhancement was an annotation at the pod that will affect all interfaces at the pod, but with multi homing we may wan to disable it at pod network but not multihoming and viceverse, that's why I added it to the ovn pod annotation where you can reference the interface by name.

An alternative would be to go back to a specific annotation but with the value being a map:
```yaml
annotations:
   "ovn.org/skip-ip-configuration-on-cni": '{"eth0": true, "eth1": false}'
```

Reply:
I don't really know why it would be per network as the enhancement has no use case described for it.

It could be as you say, a separate annotation. The annotation is pretty specific called as `skip-ip-configuration-on-cni` so it could have a default where its just
```
annotations: 
   ovn.org/skip-ip-configuration-on-cni:
```
where it applies to all attachments or to the default attachment only
and it affects specific attachments, the the value could be a json array with the attachment names
```
annotations: 
   ovn.org/skip-ip-configuration-on-cni: '["nad1". "nad2"]' 
```

Reply:
If we think there is a chance live migration could have more configuration parameters in the future, then we could just replicate the pod annotation structure but with different params
```
annotations: 
   ovn.org/skip-ip-configuration-on-cni: '{"nad1" : {"skipCNI" : true}, "nad2": {"skipCNI": true}}'
```

Reply:
> If we think there is a chance live migration could have more configuration parameters in the future, then we could just replicate the pod annotation structure but with different params
> 
> ```
> annotations: 
>    ovn.org/skip-ip-configuration-on-cni: '{"nad1" : {"skipCNI" : true}, "nad2": {"skipCNI": true}}'
> ```

Maybe we can generalize this but without mention migration

```yaml
annotations: 
   ovn.org/cni-configuration: '{"nad1" : {"skip-ip-config" : true}, "nad2": {"skip-ip-config": true}}'

```

Reply:
Would it make sense for this to be a NetConf option instead?

That Kubevirt would consume through its own NADs?

Reply:
ovn-k should be the one not setting the ip configuration from the beginning not kubevirt.

Reply:
ovn-k CNI reacts to parameters in NetConf as well and I think this option might fit there better than as a specific annotation to the pod.

So kubevirt instead of translating this configuration directly to a pod annotation, it could annotate the pod default network or other networks using NADs that include this parameter.

@maiqueb WDYT?

Reply:
I find it strange to depend on multus for something on the pod network.

I mean ... for SR-IOV I guess we have to, because that already depends on multus. 

But here, we really don't.

Reply:
We depend on multus to customize a pod's network configuration outside of what you can specify on a default cluster wide Netconf or to set the pod on the host network, it's not only about SR-IOV...it might be our only use case tight now but it's not the same thing.

This would be the first network configuration parameter that would break that rule and I would say that feels weird as well for many other reasons.

And we depend on multus to support multi-net anyway.

Do we have a use case without multus?

Up to you guys, but a CNI specific configuration parameter feels like it should be in Netconf and not in a pod annotation.

Reply:
I guess that what bothers me most here is:

- We keep providing user configuration toggles through annotations, which is not really proper.
- This configuration would be out there in the open for everyone to use. Would it had it's proper configuration place, we could make it more an internal thing.

I am not really sure how we are placing this feature outside of kubevirt/openshift world which would have multus anyway so I can't really say.

If you think an annotation on the pods is the best or only way, then I am ok with it.

Reply:
Let me be clear: if we can find a simple way to do this without the annotation, I'm all for it.

I just think passing it via CNI args via multus to the default network config is a hard sell. 

... but as you say we do that already to override the default network config. I mentioned SR-IOV because I thought that's the "real" use case for overriding the default cluster network. 

A CRD would probably be the right way to achieve this ... Maybe we could go forward with something *now* and create an epic to properly expose this configuration knob ?

Reply:
@maiqueb for multi-homing a flag at NAD is the way to go right, maybe this is not that bad ? at the end we have multus at openshift already and at hypershift we can create/delete whatever resources that we need, it means more work with the integration but is not that bad.

Today I have some meeting with hypershift team I will ask them and we can decide.

Reply:
Right; after giving this another round of thinking, I agree with CNI args.

The motivation for having this dict thing is to prepare the ground for when this skip IP configuration thing will be re-used for multi-networks - which requires multus.

I'm OK if you prefer to pass this configuration using CNI args / runtime configuration via multus.
@jcaamano / @qinqon

Reply:
After talking with hypershift kubevirt team, looks like they are ok using multus for this since the hypershift operator can construct whatever we want, so if everyone is ok here I will implement this as NetConf thing @trozet @martinkennelly are you fine with this ?

Reply:
@jcaamano @maiqueb @trozet, I have an alternative to multus:

The problem with setting ovn pod annotation is that is suppose to be own by ovn-kubernetes so no more than one controller update it, one thing we can do is that if ovn-k detects that the pod is live migratable it add this flag at the same time it allos put the allocated IP then the CNI will read it and do as expected, this way ovn-k controller continue to be the only owner and users are able to have the feature with live migration.

For future multi-homing issue the NAD controller will read the flag from the current nad and add it to the pod annotations so ovn-k continue to be owner.

What do you think ?

Reply:
That looks ok to me for live migration. As I said, I am completely unaware of the multi-homing use case but if a NetConf parameter is the solution there, then I would still consider using NetConf for live migration as well because otherwise we end up with the same parameter in two different places and while we might be able to deal with it, it still looks worse than only having it at one place.

Anyway, this idea does get rid of the main problem so not against.

Reply:
Makes sense to me. 

As I see it, going forward, whenever we implement the IPAM (sticky IPs) story, (on multi-homing) we could set the skip IP config annotation whenever the NAD does **not** have the subnet defined.

Reply:
Now that I think of if we make Default NetConf skipIPConfig to be true if pod is live migratable we don't need the ovn pod annotation thing and still future multi homing can pass the attribute at NetConf